### PR TITLE
[tcgc] add `@clientInitialization` decorator

### LINKED ISSUE #914: TCGC: @methodLocation decorator
Is it possible to move parameters that persist on each operation to the client instead with a decorator of some sort? In EventGrid this shows up in the case of
 
client()
publish(topic_name, events)
receive(topic_name, subscription_name, max_events)
ack(topic_name, subscription_name, lock_tokens)
 
to 
 
client(topic_name, subscription_name)
publish(events)
receive(max_events)
ack(lock_tokens)

typespec spec: https://github.com/Azure/azure-rest-api-specs/blob/9df71d5a717e4ed5e6728e7e6ba2fead60f62243/specification/eventgrid/Azure.Messaging.EventGrid/main.tsp#L261

### PR DESCRIPTION
**Pull request by iscai-msft**
fixes #914

### CODE REVIEW

File: packages/typespec-client-generator-core/test/decorators.test.ts [hunk 1]
```diff
@@ -4743,4 +4743,256 @@ describe("typespec-client-generator-core: decorators", () => {
...
+
+      const methods = client.methods;
+      strictEqual(methods.length, 2);
+      const download = methods[0];
+      strictEqual(download.name, "download");
+      strictEqual(download.kind, "basic");
+      strictEqual(download.parameters.length, 2);
+
+      const blobNameParam = download.parameters.find((x) => x.name === "blobName");
+      ok(blobNameParam);
```
Comment (chunyu3) [inline packages/typespec-client-generator-core/test/decorators.test.ts:4911]: parameter `blobName` is bump to client, so parameter `blobName` is not the paramter of the `download` method

**Review by tadelesh — COMMENTED** [review]
three concerns:
1. can we have a test with `@client` and `@operationGroup`? 
2. i'm wondering if we need to add a lint to disallow promoting method param to client if it is not in all client's operation?
3. with this decorator, is it possible to remove the hard code for promoting `subscriptionId` cause it is not already client parameter in brownfield?

**Comment by iscai-msft** [comment]
> three concerns:
> 
> 1. can we have a test with `@client` and `@operationGroup`?
> 2. i'm wondering if we need to add a lint to disallow promoting method param to client if it is not in all client's operation?
> 3. with this decorator, is it possible to remove the hard code for promoting `subscriptionId` cause it is not already client parameter in brownfield?

1. Yep, let me add this test
2. I think this is a fair linter-rule to add for unbranded, but it's complicated because of ARM's behavior. Currently the `list` operations don't need `subscriptionId` but all ARM sdks still have `subscriptionId` as a client-level param
3. Yeah that is down the line, where ARM can auto-apply the `@clientInitialization` decorators for `subscriptionId`

**Review by chunyu3 — COMMENTED** [review]
When the parameter `blobName` is bumped to client-level, there will no `blobName` in method (SdkMethod), but it will still in Operation (method.Operation). Right?
But we need to know that `blobName` is bumped to client-level, so that when we handle this parameter we will find it in Client. It is better to have a flag to indicate that the operation parameter is moved to client, such as Location="client"

**Comment by iscai-msft** [comment]
@chunyu3 it will have property `.onClient` set to `true`

File: packages/typespec-client-generator-core/test/decorators.test.ts [hunk 2]
```diff
@@ -3785,4 +3785,282 @@ describe("typespec-client-generator-core: decorators", () => {
...
+      strictEqual(blobNameParam.onClient, true);
+      strictEqual(blobNameParam.optional, false);
+      strictEqual(blobNameParam.kind, "method");
+
+      const blobClient = getBlobClient.response;
+
+      strictEqual(blobClient.kind, "client");
+      strictEqual(blobClient.name, "BlobClient");
+      strictEqual(blobClient.initialization.access, "public");
```
Comment (iscai-msft) [inline packages/typespec-client-generator-core/test/decorators.test.ts:3891]: my understanding was that `interface`s mapped to `operation groups` when no decorators were set, but it seems that the initialization is still public. That's totally fine, I jsut want to clarify my understanding @tadelesh. That's why I updated my code so that we only throw an error if there are new client params introduced for ogs explicitly titled with `@operationGroup`

Comment (tadelesh) [reply packages/typespec-client-generator-core/test/decorators.test.ts:3891]: i think whether explicitly set or not, the behavior should be same. why op from interface infer could be public and `@operationGroup` be internal?

File: packages/typespec-client-generator-core/generated-defs/Azure.ClientGenerator.Core.ts [hunk 3]
```diff
@@ -433,6 +433,36 @@ export type UseSystemTextJsonConverterDecorator = (
...
+ *
+ * @@clientInitialization(MyService, MyServiceClientOptions)
+ * // The generated client will have `blobName` on it. We will also
+ * // elevate the existing `blobName` parameter to the client level.
+ * ```
+ */
+export type ClientInitializationDecorator = (
+  context: DecoratorContext,
+  target: Namespace | Interface,
+  options: Model,
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/generated-defs/Azure.ClientGenerator.Core.ts:462]: is it possible that we have a consolidated decorator signature like `@override` if we want to make this decorator flexible like redefine the initialization?

File: packages/typespec-client-generator-core/test/decorators.test.ts (hunk 2 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/test/decorators.test.ts:3891]: i'm not sure, that's the current behavior though, so I think there's something wrong with current logic. Let me take a look today

Comment (weidongxu-microsoft) [reply packages/typespec-client-generator-core/test/decorators.test.ts:3891]: I kind of agree with Chenjie, that they be same.

A different scenario would be namespace nested under namespace (which is the subclient? -- do we have similar way to specify this with explicit `@client`)
```
namespace Client {
  namespace SubClient {
  }
}
```

File: docs/libraries/typespec-client-generator-core/reference/decorators.md [hunk 4]
```diff
@@ -378,6 +417,46 @@ op myOperationCustomization(params: Params): void;
...
+op upload(blobName: string): void;
+
+// client.tsp
+namespace MyCustomizations;
+model MyServiceClientOptions {
+  blob: string;
+}
+
+@@clientInitialization(MyService, MyServiceClientOptions)
+@@paramAlias(MyServiceClientOptions.blob, "blobName")
```
Comment (weidongxu-microsoft) [inline docs/libraries/typespec-client-generator-core/reference/decorators.md:454]: Does this syntac only cover `blobName` in op (like `op upload(blobName: string)`), or both this and `blob` (e.g. `op download(blob: string`?)

From your test, it appears to be latter?

File: packages/typespec-client-generator-core/test/decorators.test.ts [hunk 5]
```diff
@@ -3785,4 +3785,334 @@ describe("typespec-client-generator-core: decorators", () => {
...
+      const methods = client.methods;
+      strictEqual(methods.length, 2);
+      const download = methods[0];
+      strictEqual(download.name, "download");
+      strictEqual(download.kind, "basic");
+      strictEqual(download.parameters.length, 1);
+
+      const downloadBlobNameParam = download.parameters.find((x) => x.name === "blobName");
+      ok(downloadBlobNameParam);
+      strictEqual(downloadBlobNameParam.onClient, true);
```
Comment (weidongxu-microsoft) [inline packages/typespec-client-generator-core/test/decorators.test.ts:4106]: From this test, it appears that TCGC automatically rename the `blob` in `download` to `blobName`, according to `@paramAlias`?

So emitter would only need to map it by name.

Is it correct understanding?

**Review by weidongxu-microsoft — APPROVED** [review]
LGTM, having a few query though.

File: packages/typespec-client-generator-core/test/decorators.test.ts [hunk 6]
```diff
@@ -3785,4 +3785,334 @@ describe("typespec-client-generator-core: decorators", () => {
...
+        op download(@path blobName: string): void;
+        `,
+        `
+        namespace MyCustomizations;
+
+        model MyClientInitialization {
+          blobName: string;
+        }
+
+        @@clientInitialization(MyService, MyCustomizations.MyClientInitialization);
```
Comment (haolingdong-msft) [inline packages/typespec-client-generator-core/test/decorators.test.ts:3804]: So from TCGC's response perspective, there is not concept of `clientInitialization`, emitter just need to process the `SdkParameter` when process each operation. Does my understanding correct?

File: packages/typespec-client-generator-core/test/decorators.test.ts (hunk 5 above)
Comment (weidongxu-microsoft) [reply packages/typespec-client-generator-core/test/decorators.test.ts:4106]: We may need to be extra careful, if service write e.g.

```
model RequestBody {
  blobName: string;
  otherProperties: ...;
}

op action(@path blob: string, ...RequestBody);
```

File: packages/typespec-client-generator-core/test/decorators.test.ts (hunk 6 above)
Comment (weidongxu-microsoft) [reply packages/typespec-client-generator-core/test/decorators.test.ts:3804]:
```
    initialization: diagnostics.pipe(getSdkInitializationType(context, client)),
```

File: packages/typespec-client-generator-core/generated-defs/Azure.ClientGenerator.Core.ts (hunk 3 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/generated-defs/Azure.ClientGenerator.Core.ts:462]: I'm not quite sure what you're suggesting, are you suggesting we use `@override` instead of `@clientInitialization`? If so, in some ways they're similar, but I feel since there is no original "client initialization" that is defined in the tsp, using the decorator `@override` wouldn't make sense to users

File: docs/libraries/typespec-client-generator-core/reference/decorators.md (hunk 4 above)
Comment (iscai-msft) [reply docs/libraries/typespec-client-generator-core/reference/decorators.md:454]: Both! It's saying `blob` and `blobName` in paramters refer to the same thing

File: packages/typespec-client-generator-core/test/decorators.test.ts (hunk 5 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/test/decorators.test.ts:4106]: This is because we have a bug where we're still adding some client params to the method signature. That fix has been merged and I'm going to fix the test in this PR. Good catch!

File: packages/typespec-client-generator-core/test/decorators.test.ts (hunk 6 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/test/decorators.test.ts:3804]: @haolingdong-msft the model in `@clientInitialization` gets added as the `initialization` property on the `SdkClientType`. They are treated as method parameters, so we don't have a concept of them being query / body etc: that we can find out in `method.operation` from the params being mapped back to them

File: packages/typespec-client-generator-core/src/decorators.ts [hunk 7]
```diff
@@ -989,3 +1006,68 @@ export const $useSystemTextJsonConverter: DecoratorFunction = (
...
+  if (hasExplicitClientOrOperationGroup(context)) {
+    isOg = context.program.stateMap(operationGroupKey).get(target) !== undefined;
+  }
+  if (isOg) {
+    if (
+      target.namespace &&
+      context.program.stateMap(clientInitializationKey).get(target.namespace) &&
+      context.program.stateMap(clientInitializationKey).get(target.namespace)[AllScopes] === options
+    ) {
+      setScopedDecoratorData(context, $override, clientInitializationKey, target, options, scope);
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/src/decorators.ts:1028]: is `$override` a typo?

File: packages/typespec-client-generator-core/src/decorators.ts [hunk 8]
```diff
@@ -989,3 +1006,68 @@ export const $useSystemTextJsonConverter: DecoratorFunction = (
...
+    } else {
+      reportDiagnostic(context.program, {
+        code: "assigning-public-params-to-internal-client",
+        format: { name: target.name },
+        target: context.decoratorTarget,
+      });
+    }
+  }
+  setScopedDecoratorData(context, $override, clientInitializationKey, target, options, scope);
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/src/decorators.ts:1037]: add `else` to prevent set after diagnostic?

File: packages/typespec-client-generator-core/src/decorators.ts [hunk 9]
```diff
@@ -989,3 +1006,68 @@ export const $useSystemTextJsonConverter: DecoratorFunction = (
...
+  }
+  if (isOg) {
+    if (
+      target.namespace &&
+      context.program.stateMap(clientInitializationKey).get(target.namespace) &&
+      context.program.stateMap(clientInitializationKey).get(target.namespace)[AllScopes] === options
+    ) {
+      setScopedDecoratorData(context, $override, clientInitializationKey, target, options, scope);
+    } else {
+      reportDiagnostic(context.program, {
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/src/decorators.ts:1030]: for .net, they could support sub client with param, and they will generate such api for face. so, do we really need to add this check?

File: packages/typespec-client-generator-core/src/http.ts [hunk 10]
```diff
@@ -526,7 +546,7 @@ export function getCorrespondingMethodParams(
       );
       return diagnostics.wrap([]);
     }
-    return diagnostics.wrap([context.__subscriptionIdParameter]);
+    return diagnostics.wrap(subId ? [subId] : []);
   }
 
   // to see if the service parameter is a method parameter or a property of a method parameter
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/src/http.ts:552]: from this logic, we could not pop a method's model param's property to client, right?

**Review by tadelesh — COMMENTED** [review]
can we add a test of redefining the client hierarchy in client.tsp and then add `@clientInitialization` to the new defined client? it seems will work well from the code logic, but i'm not sure if there is any corner case.

**Review by tadelesh — APPROVED** [review]
overall lgtm. one thing is could you also check if we could solve this: https://github.com/Azure/typespec-azure/issues/1391 in this pr?

File: packages/typespec-client-generator-core/generated-defs/Azure.ClientGenerator.Core.ts [hunk 11]
```diff
@@ -433,6 +433,67 @@ export type UseSystemTextJsonConverterDecorator = (
...
+ * namespace MyCustomizations;
+ * model MyServiceClientOptions {
+ *   blob: string;
+ * }
+ *
+ * @@clientInitialization(MyService, MyServiceClientOptions)
+ * @@paramAlias(MyServiceClientOptions.blob, "blobName")
+ *
+ * // The generated client will have `blobName` on it. We will also
+ * // elevate the existing `blob` parameter to the client level.
```
[the remark this answers is not in the document]
Comment (iscai-msft) [reply packages/typespec-client-generator-core/generated-defs/Azure.ClientGenerator.Core.ts:487]: yep thanks for the catch!

File: packages/typespec-client-generator-core/src/decorators.ts [hunk 12]
```diff
@@ -87,13 +92,18 @@ function getScopedDecoratorData(
   return retval[AllScopes]; // in this case it applies to all languages
 }
 
-function listScopedDecoratorData(context: TCGCContext, key: symbol): any[] {
+function listScopedDecoratorData(
+  context: { program: Program },
+  key: symbol,
+  emitterName?: string
+): any[] {
```
[the remark this answers is not in the document]
Comment (iscai-msft) [reply packages/typespec-client-generator-core/src/decorators.ts:99]: let me roll back this change, I needed it at one point, but you're right we don't need to modify this code

File: packages/typespec-client-generator-core/src/decorators.ts (hunk 9 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/src/decorators.ts:1030]: let me go ahead and get rid of this check. This is explicitly for operation groups, not initializable subclients, but it is making the code more complicated than it needs to be, so let me roll this back

File: packages/typespec-client-generator-core/src/http.ts (hunk 10 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/src/http.ts:552]: yes we can, this is returning corresponding method params, which in this case we're returning a client param
